validator/client/src/test/java/tech/pegasys/teku/validator/client/StableSubnetSubscriberTest.java

```diff
@@ -0,0 +1,195 @@
+    verify(validatorApiChannel).subscribeToPersistentSubnets(subnetSubcriptions.capture());
+    assertThat(subnetSubcriptions.getValue()).hasSize(Constants.ATTESTATION_SUBNET_COUNT);
+    assertSubnetsAreDistinct(subnetSubcriptions.getValue());
+    assertUnsubscribeSlotsAreInBound(subnetSubcriptions.getValue(), valueOf(15));
+  }
+
+  @Test
+  void shouldStaySubscribedToAllSubnetsEvenIfValidatorNumberIsDecreased() {
+    ValidatorApiChannel validatorApiChannel = mock(ValidatorApiChannel.class);
+    StableSubnetSubscriber stableSubnetSubscriber =
+        new StableSubnetSubscriber(
+            validatorApiChannel, new Random(), Constants.ATTESTATION_SUBNET_COUNT + 8);
+
+    stableSubnetSubscriber.onSlot(ZERO);
+
+    ArgumentCaptor<Set<SubnetSubscription>> subnetSubcriptions = ArgumentCaptor.forClass(Set.class);
+    verify(validatorApiChannel).subscribeToPersistentSubnets(subnetSubcriptions.capture());
+    assertSubnetsAreDistinct(subnetSubcriptions.getValue());
+    assertThat(subnetSubcriptions.getValue()).hasSize(Constants.ATTESTATION_SUBNET_COUNT);
+
+    stableSubnetSubscriber.updateValidatorCount(Constants.ATTESTATION_SUBNET_COUNT);
+    stableSubnetSubscriber.onSlot(valueOf(2));
+
+    verifyNoMoreInteractions(validatorApiChannel);
+  }
+
+  @Test
+  void shouldReplaceExpiredSubscriptionsWithNewOnes() {
+    ValidatorApiChannel validatorApiChannel = mock(ValidatorApiChannel.class);
+
+    StableSubnetSubscriber stableSubnetSubscriber =
+        new StableSubnetSubscriber(validatorApiChannel, new Random(), 2);
+
+    stableSubnetSubscriber.onSlot(valueOf(0));
+
+    ArgumentCaptor<Set<SubnetSubscription>> firstSubscriptionUpdate =
+        ArgumentCaptor.forClass(Set.class);
+    ArgumentCaptor<Set<SubnetSubscription>> secondSubscriptionUpdate =
+        ArgumentCaptor.forClass(Set.class);
+
+    verify(validatorApiChannel).subscribeToPersistentSubnets(firstSubscriptionUpdate.capture());
+
+    assertThat(firstSubscriptionUpdate.getValue()).hasSize(2);
+    assertThat(firstSubscriptionUpdate.getValue()).hasSize(2);
+
+    UnsignedLong firstUnsubscriptionSlot =
+        firstSubscriptionUpdate.getValue().stream().findFirst().get().getUnsubscriptionSlot();
+
+    stableSubnetSubscriber.onSlot(firstUnsubscriptionSlot.minus(UnsignedLong.ONE));
+
+    verifyNoMoreInteractions(validatorApiChannel);
+    stableSubnetSubscriber.onSlot(firstUnsubscriptionSlot);
+
+    verify(validatorApiChannel, times(2))
+        .subscribeToPersistentSubnets(secondSubscriptionUpdate.capture());
+
+    UnsignedLong secondUnsubscriptionSlot =
+        secondSubscriptionUpdate.getValue().stream().findFirst().get().getUnsubscriptionSlot();
+
+    assertThat(firstUnsubscriptionSlot).isNotEqualByComparingTo(secondUnsubscriptionSlot);
+    // Can only verify unsubscription slot have changed and not the subnet id,
+    // since subnet id can randomly be chosen the same
+  }
+
+  private void assertUnsubscribeSlotsAreInBound(
```

Comment:
Since we're dealing with randomness, might be nice to add a test that generates a large number of subscriptions and checks that they're all correct (in bounds and distinct).

Reply:
Added.